### PR TITLE
Support multiple configs for each hook types and specifying the hooks in the alarm rule.

### PR DESCRIPTION
**Pull request by wankai123**
- [ ] If this is non-trivial feature, paste the links/URLs to the design doc.
- [X] Update the documentation to include this new feature.
- [X] Tests(including UT, IT, E2E) are added to verify the new feature.
- [ ] If it's UI related, attach the screenshots below.

- [ ] If this pull request closes/resolves/fixes an existing issue, replace the issue number. Closes #<issue number>.
- [X] Update the [`CHANGES` log](https://github.com/apache/skywalking/blob/master/docs/en/changes/changes.md).

### CODE REVIEW

File: docs/en/setup/backend/backend-alarm.md [hunk 1]
```diff
@@ -223,85 +271,95 @@ message KeyStringValuePair {
...
+  default:
+    isGlobal: true
+    textTemplate: |-
     {
       "type": "section",
       "text": {
         "type": "mrkdwn",
         "text": ":alarm_clock: *Apache Skywalking Alarm* \n **%s**."
       }
     }
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-alarm.md:289]:
```suggestion
      {
        "type": "section",
        "text": {
          "type": "mrkdwn",
          "text": ":alarm_clock: *Apache Skywalking Alarm* \n **%s**."
        }
      }
```

File: docs/en/setup/backend/backend-alarm.md [hunk 2]
```diff
@@ -223,85 +271,95 @@ message KeyStringValuePair {
...
-  textTemplate: |-
+  default:
+    isGlobal: true
+    textTemplate: |-
     {
       "msgtype": "text",
       "text": {
         "content": "Apache SkyWalking Alarm: \n %s."
       }
     }
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-alarm.md:307]:
```suggestion
      {
        "msgtype": "text",
        "text": {
          "content": "Apache SkyWalking Alarm: \n %s."
        }
      }
```

File: docs/en/setup/backend/backend-alarm.md [hunk 3]
```diff
@@ -223,85 +271,95 @@ message KeyStringValuePair {
...
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-alarm.md:326]:
```suggestion
      {
        "msgtype": "text",
        "text": {
          "content": "Apache SkyWalking Alarm: \n %s."
        }
      }
```

File: oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/AlarmHookSettings.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    @Getter
+    private final AlarmHooksType type;
+    @Getter
+    private final boolean isGlobal;
+
+   public AlarmHookSettings(String name, AlarmHooksType type, boolean isGlobal) {
+        this.name = name;
+        this.type = type;
+        this.isGlobal = isGlobal;
+    }
```
Comment (kezhenxu94) [inline oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/AlarmHookSettings.java:35]: Replace with `@RequiredArgsConstructor`

File: docs/en/setup/backend/backend-alarm.md [hunk 5]
```diff
@@ -123,13 +129,20 @@ rules:
     silence-period: 5
     message: The request number of entity {name} non-200 status is more than expected.
     only-as-condition: false
+    specific-hooks:
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-alarm.md:132]:
```suggestion
   hooks:
```

I think hooks is a good name already.

File: docs/en/setup/backend/backend-alarm.md [hunk 6]
```diff
@@ -152,7 +165,42 @@ Currently, metrics from the **Service**, **Service Instance**, **Endpoint**, **S
...
+## Hooks
+Hooks are a way to send alarm messages to the outside world. SkyWalking supports multiple hooks of the same type, each hook can support different configurations. 
+For example, you can configure two Slack hooks, one named `default` and set `isGlobal: true` means this hook will apply on all `Alarm Rules` **without config** `specific-hooks`.
+Another named `custom1` will only apply on the `Alarm Rules` which **with config** `specific-hooks` and include the name `slackHooks.custom1`.
+
+```yaml
+hooks:
+  slackHooks:
+    default:
+      isGlobal: true # if true, this hook will apply on all rules, unless a rule has its own specific hook.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-alarm.md:177]: Will mutiple global rules conflict? Or we will pick one randomly?

File: docs/en/setup/backend/backend-alarm.md [hunk 7]
```diff
@@ -368,4 +430,5 @@ the sliding window will be destroyed and re-created, causing the Alarm of this s
 | Count                | count                      | int            |                    |
 | Only as condition    | only-as-condition          | boolean        |                    |
 | Silence period       | silence-period             | int            |                    |
-| Message              | message                    | string         |                    |
\ No newline at end of file
+| Message              | message                    | string         |                    |
+| Specific Hooks       | specific-hooks             | string array   |                    |
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-alarm.md:434]: I think we don't specific hook name, as users are set the server side as known.

File: docs/en/setup/backend/backend-alarm.md (hunk 6 above)
Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md:177]: support multiple global hooks in one type, if they have different names

Comment (wu-sheng) [reply docs/en/setup/backend/backend-alarm.md:177]: So, how would we react when conflicts? Are we still forwarding all as users say so?

File: oap-server/server-starter/src/main/resources/alarm-settings.yml [hunk 8]
```diff
@@ -41,81 +41,128 @@ rules:
...
 
-pagerDutyHooks:
-  textTemplate: "Apache SkyWalking Alarm: \n %s."
-  integrationKeys:
-#    # you can find your integration key(s) on the Events API V2 integration page for your PagerDuty service(s).
-#    # (you may need to create an Events API V2 integration for your PagerDuty service if you don't have one yet)
-#    # below are dummy keys that should be replaced with your own integration keys.
-#    - dummy_key
-#    - dummy_key2
+hooks:
```
Comment (kezhenxu94) [inline oap-server/server-starter/src/main/resources/alarm-settings.yml:54]: Why do we want to add a section level `hooks` here? The following sections all have a `**hooks` suffix already and this looks pretty verbose

File: docs/en/setup/backend/backend-alarm.md (hunk 6 above)
Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md:177]: If use the `yaml` settings, we can't config the same name under one item. such as we can't define 2 `default` name under slackHooks

Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md:177]: Syntactic restriction

File: oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/discord/DiscordHookCallback.java [hunk 9]
```diff
@@ -43,17 +44,28 @@ public class DiscordHookCallback extends HttpAlarmCallback {
...
+
+        Map<String, List<AlarmMessage>> groupedMessages = groupMessagesByHook(alarmMessages);
+        for (Map.Entry<String, List<AlarmMessage>> entry : groupedMessages.entrySet()) {
+            var hookName = entry.getKey();
+            var messages = entry.getValue();
+            var setting = settingsMap.get(hookName);
+            if (setting == null || CollectionUtils.isEmpty(setting.getWebhooks()) || CollectionUtils.isEmpty(
+                messages)) {
+                continue;
+            }
```
Comment (kezhenxu94) [inline oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/discord/DiscordHookCallback.java:60]: You can't simply skip if `settings == null`, `settings` being null has high possibility that the users configure a wrong name that doesn't exist at all (e.g. typo), you should check the hook names exist when reading the configs and bail out as a startup exception.

File: docs/en/setup/backend/backend-alarm.md (hunk 6 above)
Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md:177]: The `default` is just a name and has no special meaning, users can define it themself.

File: oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/webhook/WebhookCallback.java [hunk 10]
```diff
@@ -39,13 +41,24 @@ public class WebhookCallback extends HttpAlarmCallback {
...
             return;
         }
 
-        for (String url : alarmRulesWatcher.getWebHooks()) {
-            post(URI.create(url), gson.toJson(alarmMessage), Map.of());
+        Map<String, List<AlarmMessage>> groupedMessages = groupMessagesByHook(alarmMessages);
+        for (Map.Entry<String, List<AlarmMessage>> entry : groupedMessages.entrySet()) {
+            var hookName = entry.getKey();
+            var messages = entry.getValue();
+            var setting = settingsMap.get(hookName);
```
Comment (kezhenxu94) [inline oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/webhook/WebhookCallback.java:54]: The `hookName` from the `groupMessagesByHook(alarmMessages);` is something like `slackHooks.custom1`, but the keys in `settingsMap` is something like `custom1`, how can you get the setting with `settingsMap.get(hookName);`? Did you process the hook name somewhere I missed?

And the tests only covers global webhooks, please test the rule-level web hooks

**Review by kezhenxu94 — CHANGES_REQUESTED** [review]

File: oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/webhook/WebhookCallback.java (hunk 10 above)
Comment (wankai123) [reply oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/webhook/WebhookCallback.java:54]: The full name is `{hookType}.{hookName}`,  define and get from `AlarmHookSettings.java`

File: oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/discord/DiscordHookCallback.java (hunk 9 above)
Comment (wankai123) [reply oap-server/server-alarm-plugin/src/main/java/org/apache/skywalking/oap/server/core/alarm/provider/discord/DiscordHookCallback.java:60]: Added

File: docs/en/setup/backend/backend-alarm.md [hunk 11]
```diff
@@ -152,7 +165,42 @@ Currently, metrics from the **Service**, **Service Instance**, **Endpoint**, **S
...
 
-## Webhook
+## Hooks
+Hooks are a way to send alarm messages to the outside world. SkyWalking supports multiple hooks of the same type, each hook can support different configurations. 
+For example, you can configure two Slack hooks, one named `default` and set `is-global: true` means this hook will apply on all `Alarm Rules` **without config** `hooks`.
+Another named `custom1` will only apply on the `Alarm Rules` which **with config** `hooks` and include the name `slack.custom1`.
+
+```yaml
+hooks:
+  slackHooks:
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-alarm.md:175]:
```suggestion
  slack:
```

File: docs/en/setup/backend/backend-alarm.md [hunk 12]
```diff
@@ -223,120 +271,134 @@ message KeyStringValuePair {
...
+  default:
+    is-global: true
+    text-template: |-
     {
       "msg_type": "text",
       "content": {
         "text": "Apache SkyWalking Alarm: \n %s."
       },
       "ats":"feishu_user_id_1,feishu_user_id_2"
     }
```
Comment (kezhenxu94) [inline docs/en/setup/backend/backend-alarm.md:348]:
```suggestion
      {
        "msg_type": "text",
        "content": {
          "text": "Apache SkyWalking Alarm: \n %s."
        },
        "ats":"feishu_user_id_1,feishu_user_id_2"
      }
```

**Review by kezhenxu94 — APPROVED** [review]

**Review by kezhenxu94 — DISMISSED** [review]

File: docs/en/setup/backend/backend-alarm.md [hunk 13]
```diff
@@ -152,7 +165,42 @@ Currently, metrics from the **Service**, **Service Instance**, **Endpoint**, **S
...
+## Hooks
+Hooks are a way to send alarm messages to the outside world. SkyWalking supports multiple hooks of the same type, each hook can support different configurations. 
+For example, you can configure two Slack hooks, one named `default` and set `is-global: true` means this hook will apply on all `Alarm Rules` **without config** `hooks`.
+Another named `custom1` will only apply on the `Alarm Rules` which **with config** `hooks` and include the name `slack.custom1`.
+
+```yaml
+hooks:
+  slack:
+    default:
+      is-global: true # if true, this hook will apply on all rules, unless a rule has its own specific hook.
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-alarm.md:177]: > unless a rule has its own specific hook.

So, if there are `hooks` specified in the rules, the `is-global` is not working.
Am I right? If so, the name `global` seems not proper, it is more like a default rule. We should consider not to set `default` as name in the demo, and change this to `is-default : true`.

Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md:177]: > So, if there are hooks specified in the rules, the is-global is not working. Am I right?

yes

> If so, the name global seems not proper, it is more like a default rule. We should consider not to set default as name in the demo, and change this to is-default : true

I think `is-global` means you can define more than one global hook in one type. `is-default` mostly means only one?

Comment (wu-sheng) [reply docs/en/setup/backend/backend-alarm.md:177]: But `global` seems not to be able to be overridden.

Comment (wankai123) [reply docs/en/setup/backend/backend-alarm.md]: make sense, changed

File: docs/en/setup/backend/backend-alarm.md [hunk 14]
```diff
@@ -152,7 +165,43 @@ Currently, metrics from the **Service**, **Service Instance**, **Endpoint**, **S
...
-## Webhook
+## Hooks
+Hooks are a way to send alarm messages to the outside world. SkyWalking supports multiple hooks of the same type, each hook can support different configurations. 
+For example, you can configure two Slack hooks, one named `default` and set `is-default: true` means this hook will apply on all `Alarm Rules` **without config** `hooks`.
+Another named `custom1` will only apply on the `Alarm Rules` which **with config** `hooks` and include the name `slack.custom1`.
+
+```yaml
+hooks:
+  slack:
+    default:
```
Comment (wu-sheng) [inline docs/en/setup/backend/backend-alarm.md:176]:
```suggestion
    # default here is just a name, set the field default: true if this notification hook is expected to be default globally.
    default:
```

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**